### ISSUE TITLE
[UX] URL alias patterns page: "Browse available tokens ..." link gets out of viewport if too many vocabularies/content types

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#3187]
I don't have an answer for this, but it happens when there are too many content types and/or vocabularies on a site.

STR:

1. Either use the devel module to automatically generate vocabularies, or do it manually ...how many depends on your screen size (I have created 20).
2. Scroll to add a pattern to the last vocabulary in the list -> click on the respective field to gain focus
3. If you want to use the token browser in order to enter tokens, you need to scroll back up to get to the "Browse available tokens for ..." link...
4. Once you have clicked the link and the token browser dialog has opened, you need to now scroll back down in order to bring the field you previously selected in step 2 within view.

...if the list of vocabularies is too long, you will need to be dragging the token browser dialog along, so there will be a lot of scroll -> drag -> repeat ...until you reach the vocabulary.

We could use some fancy JS in order to keep the token browser dialog within view as the user scrolls, but this would solve only part of the problem. So the question is how could we keep the "Browse available tokens for ..." link close to the field that the user intends to use it for?

---
PoC PR by @klonos (credits to @serundeputy): https://github.com/backdrop/backdrop/pull/2239

(I am looking for issues to mark as accessibility issues for an upcoming Sprint. It's not clear to me that this one would qualify. Please post a comment if you disagree. - @stpaultim 08/17/2021).

**Comment by docwilmot** [comment]
How about autocomplete for token enabled fields? You type `[` and it suggests all top-level tokens, then you continue with `[nod...` it gives the `[node:token]` ones etc.

**Comment by klonos** [comment]
Yeah, that could work @docwilmot ...Rules-like 😉...I see how I could get myself used to that. ...the only things I see us loosing this way is the token descriptions (clicking the "more" link), and browsing the tokens via point and click (without having to know what you should be looking for).

I was exploring the option of converting each one of these fieldsets into tables, with the token pattern fields being the rows. We could then stick the token browser link in the table header, which would inherit the stickiness.

Hmm.

**Comment by serundeputy** [comment]
I think a token browser link on each field is the way to go:

<img width="893" alt="screen shot 2018-07-05 at 12 08 22 pm" src="https://user-images.githubusercontent.com/6763671/42334826-5e5603be-804c-11e8-9215-67ed1b2c5a6e.png">


some codez in `path.admin.inc`:

```php
   // Display the user documentation of placeholders supported by this path
    // type as a description on the last pattern.
    $variables = array(
      'token_types' => $token_types,
      'text' => t(
          '&&',
          array('@type' => $path_info['label']
        )
      ),
      'title' => t('tokens'),
    );
    // Prompt for the default URL alias pattern for this path type.
    $variable = $path_type . '_pattern';
    $token_browser_link =
      '<span class="token-browser-link">' .
        theme('token_tree_link', $variables) .
      '</span>';
    $form[$path_type][$variable] = array(
      '#type' => 'textfield',
      '#title' => $pattern_description,
      '#default_value' => $config->get($variable),
      '#size' => 65,
      '#maxlength' => 1280,
      '#element_validate' => array('token_element_validate'),
      '#after_build' => array('token_element_validate'),
      '#token_types' => $token_types,
      '#min_tokens' => 1,
      '#parents' => array($variable),
      '#suffix' => $token_browser_link,
    );

    // Set up specialized patterns for this path type (e.g. node types).
    if (isset($path_info['pattern items'])) {
      foreach ($path_info['pattern items'] as $item_name => $item_label) {
        $variable = $path_type . '_' . $item_name . '_pattern';
        $form[$path_type][$variable] = array(
          '#type' => 'textfield',
          '#title' => $item_label,
          '#default_value' => $config->get($variable),
          '#size' => 65,
          '#maxlength' => 1280,
          '#element_validate' => array('token_element_validate'),
          '#after_build' => array('token_element_validate'),
          '#token_types' => $token_types,
          '#min_tokens' => 1,
          '#parents' => array($variable),
          '#suffix' => $token_browser_link,
        );
      }
    }

    //$form[$path_type]['#description'] = '<p>' .
    //  t('Each of the fields below support tokens.') . ' ' . theme('token_tree_link', $variables) . '</p>';
  }
```

**Comment by klonos** [comment]
Thanks @serundeputy 👍 ...I have whipped your codez into a PoC PR: https://github.com/backdrop/backdrop/pull/2239

1. Not sure where the css should live; I have for now put it in `/core/themes/seven/css/styles.css`

2. I have not altered the "Each of the fields below support tokens ..." description text. I believe that it should not be repeated for each pattern group; it belongs as an info message at the top of the page (#3155). Something along the lines of "Each of the fields below support tokens. Click on the [...] link to open the token browser."

3. I have made the link text be `[...]` in order to resemble the tokens format (could also be `[???]` or something). Upon hover, it expands to `[token browser]` (needs proper animation/toggle that does not bleed outside the filed, RTL styling, made to also expand when the field gets focus etc., ...but as I said "PoC" for now). The reason for not having the link text be the full "token browser" always is that in certain languages this might end up being a very long string, which would then occupy a big portion of each field. Also less distracting to see less text being repeated.

I have generated a ridiculous amount of vocabularies in the sandbox, and I have also left the "Browse available tokens for Content." link in place for the purpose of comparing the UX before/after.

<img width="684" alt="screen shot 2018-07-08 at 5 16 44 am" src="https://user-images.githubusercontent.com/2423362/42413936-8907c52e-826e-11e8-8f12-2a648920e211.png">
